internal/cmd/nexec.go

```diff
@@ -147,32 +149,49 @@ func execInClusterNodes(args []string) error {
 	// In distributed shell mode, TTY is forced to be disabled
 	nodeExecNoTty = true
 
+	// In case nothing is configured, all nodes will be executed concurrently
+	if nodeExecMaxParallel <= 0 {
+		nodeExecMaxParallel = len(nodes)
+	}
+
+	type task struct {
+		node   corev1.Node
+		reader io.Reader
+	}
+
 	var (
 		wg      = &sync.WaitGroup{}
+		tasks   = make(chan task, len(nodes))
 		readers = duplicateReader(os.Stdin, len(nodes))
 		output  = make(chan OutputMsg)
 		errors  = make(chan error, len(nodes))
 		printer = make(chan bool, 1)
 	)
 
-	wg.Add(len(nodes))
-	for i, node := range nodes {
-		go func(node corev1.Node, reader io.Reader) {
-			defer func() {
-				wg.Done()
-			}()
-
-			errors <- hvnr.NodeExec(
-				node,
-				nodeExecImage,
-				nodeExecTimeout,
-				command,
-				reader,
-				chanWriter("StdOut", node.Name, output),
-				chanWriter("StdErr", node.Name, output),
-				!nodeExecNoTty,
-			)
-		}(node, readers[i])
+	// Fill task queue with the list of nodes to be processed
+	for i := range nodes {
+		tasks <- task{reader: readers[i], node: nodes[i]}
+	}
+	close(tasks)
+
+	// Start n task workers to work on task queue
+	wg.Add(nodeExecMaxParallel)
```

Comment:
question: If I have 7 nodes and the value of `nodeExecMaxParallel` is 3, I was expecting to wait first for 3 routines, then another 3 and last only 1, here it looks to be waiting for 7 all the time?